language/diem-framework/modules/NetworkIdentity.move

```diff
@@ -140,25 +140,177 @@ module DiemFramework::NetworkIdentity {
+        include has_change ==> DiemTimestamp::AbortsIfNotOperating;
+        include RemoveMembersInternalEnsures<vector<u8>> {
+            old_members: prior_identities,
+            new_members: global<NetworkIdentity>(account_addr).identities,
+        };
         modifies global<NetworkIdentity>(account_addr);
-        invariant exists<NetworkIdentity>(account_addr);
-        invariant num_identities <= MAX_ADDR_IDENTITIES;
+        emits msg to handle if has_change;
+    }
+
+    // =================================================================
+    // Set operation simulation
+
+    /// Add all elements that appear in `to_add` into `members`.
+    ///
+    /// The `members` argument is essentially a set simulated by a vector, hence
+    /// the uniqueness of its elements are guaranteed, before and after the bulk
+    /// insertion. The `to_add` argument, on the other hand, does not guarantee
+    /// to be a set and hence can have duplicated elements.
+    fun add_members_internal<T: copy>(
+        members: &mut vector<T>,
+        to_add: &vector<T>,
+    ): bool {
+        let num_to_add = Vector::length(to_add);
+        let num_existing = Vector::length(members);
+
+        let i = 0;
+        while ({
+            spec {
+                invariant i <= num_to_add;
+                // the set can never reduce in size
+                invariant len(members) >= len(old(members));
+                // the current set maintains the uniqueness of the elements
+                invariant forall j in 0..len(members), k in 0..len(members): members[j] == members[k] ==> j == k;
+                // the left-split of the current set is exactly the same as the original set
+                invariant forall j in 0..len(old(members)): members[j] == old(members)[j];
+                // all elements in the the right-split of the current set is from the `to_add` vector
+                invariant forall j in len(old(members))..len(members): contains(to_add[0..i], members[j]);
+                // the current set includes everything in `to_add` we have seen so far
+                invariant forall j in 0..i: contains(members, to_add[j]);
+                // having no new members means that all elements in the `to_add` vector we have seen so far are already
+                // in the existing set, and vice versa.
+                invariant len(members) == len(old(members)) <==> (forall j in 0..i: contains(old(members), to_add[j]));
+            };
+            (i < num_to_add)
+        }) {
+            let entry = Vector::borrow(to_add, i);
+            if (!Vector::contains(members, entry)) {
+                Vector::push_back(members, *entry);
+            };
+            i = i + 1;
+        };
+
+        Vector::length(members) > num_existing
+    }
+    spec add_members_internal {
+        pragma opaque;
+        // TODO(mengxu): this is to force the prover to honor the "opaque" pragma in the ignore opaque setting
+        ensures [concrete] true;
```

Comment:
Yeah, functions with loops should probably always be opaque. For now this is a good workaround.